feat: no-std support

huitseeker opened the pull request:
Adds support for no-std.

This is a slight modernization of the miette fork by @bitwalker mentioned in #443, found at https://github.com/bitwalker/miette/compare/main...no-std

All the credit for the contribution goes to @bitwalker, I'm just doing janitorial work around the edges.

bitwalker committed:
fix: support no-std environments

huitseeker committed:
feat(no_std): Implement comprehensive no_std support

This commit finishes full no_std support to miette.

huitseeker committed:
Add CI job for no-std build verification and fix Infallible StdError implementation

- Add dedicated no-std CI job that builds for wasm32-unknown-unknown target
- Fix Infallible StdError implementation for no-std environments

The CI job validates:
1. Core no-std functionality (no default features)
2. fancy-no-syscall feature set compatible with no-std environments

cgettys-microsoft commented on miette-derive/src/code.rs:
```diff
@@ -56,13 +56,13 @@ impl Code {
                 let code = &code.as_ref()?.0;
                 Some(match fields {
                     syn::Fields::Named(_) => {
-                        quote! { Self::#ident { .. } => std::option::Option::Some(std::boxed::Box::new(#code)), }
```
why isn't Option qualified after this change? just curious

huitseeker replied:
Bare `Option::` works fine in both std and no_std because it's in the core prelude. The change from `std::option::Option` to just `Option` is actually more portable since it doesn't hardcode the std path. We could qualify it as `core::option::Option` for explicitness but it's not needed for correctness.

cgettys-microsoft commented on src/error.rs:
```diff
@@ -66,42 +77,58 @@ impl Diagnostic for MietteError {
         }
     }
 
+    #[cfg(not(feature = "std"))]
```
Curiosity: why not a different struct, instead of using one in two configs.
If the answer is "won't compile", then this is infectious / not additive, no?

cgettys-microsoft commented on src/eyreish/context.rs:
```diff
@@ -29,6 +34,7 @@ mod ext {
     }
 
     impl Diag for Report {
+        #[cfg_attr(track_caller, track_caller)]
```
is track_caller needed for the no_std implementation, or is that a separate change?

cgettys-microsoft commented on src/eyreish/error.rs:
```diff
@@ -816,8 +817,16 @@ impl AsRef<dyn StdError> for Report {
         self.as_ref()
     }
 }
+
```
WHy doesn't std::core::Borrow work in std contexts? why two impls, in other words?

cgettys-microsoft commented on src/eyreish/into_diagnostic.rs:
```diff
@@ -1,14 +1,17 @@
-use std::{error::Error, fmt::Display};
+extern crate alloc;
+
+use crate::StdError as Error;
```
why re-name StdError to Error here, but not elsewehre?

cgettys-microsoft commented on src/eyreish/kind.rs:
```diff
@@ -45,11 +45,20 @@
 //     let error = $msg;
 //     (&error).miette_kind().new(error)
 
+#[cfg(not(feature = "std"))]
```
can all these repeated declarations live in a single imported prelude instead, or nah?

cgettys-microsoft commented on src/eyreish/mod.rs:
```diff
@@ -81,18 +91,34 @@ impl Diagnostic for InstallError {}
-    let hook = HOOK.get_or_init(|| Box::new(get_default_printer)).as_ref();
+pub(crate) fn capture_handler_with_location(
+    error: &(dyn Diagnostic + 'static),
+) -> Box<dyn ReportHandler> {
```
Why the Once? can't it just be OnceLock?
I'm having a hard time following what this change is achieving, which means more comments might be needed here. I don't see why two different once's are needed.

huitseeker replied:
`OnceLock` isn't available in `no_std`, hence `spin::Once`

cgettys-microsoft commented on src/eyreish/wrapper.rs:
```diff
@@ -107,16 +110,6 @@ impl StdError for BoxedError {
     fn source(&self) -> Option<&(dyn StdError + 'static)> {
         self.0.source()
     }
-
```
This is a breaking change, no?

huitseeker replied:
These [have been deprecated since rust 1.42/1.33](https://github.com/rust-lang/rust/blob/86a49fd71fecd25b0fd20247db0ba95eeceaba28/library/core/src/error.rs).

cgettys-microsoft commented on src/eyreish/wrapper.rs:
```diff
@@ -216,8 +209,16 @@ impl<C> StdError for WithSourceCode<Report, C> {
+        boxed::Box,
+        string::{String, ToString},
+    };
+    #[cfg(feature = "fancy")]
```
shouldn't many of these still be alloc:: qualified types instead, in case someone wants to make fancy no-std too in future?

huitseeker replied:
Fair point for future-proofing. Right now the fancy features all require std (terminal detection, env vars, etc.) so these std imports are fine. If someone wanted `no_std_fancy` later they'd need to rework the syscall detection anyway. Left it as-is to keep the diff smaller but happy to change if you feel strongly.

cgettys-microsoft commented on src/handler.rs:
```diff
@@ -491,21 +507,27 @@ mod syscall {
+            if #[cfg(feature = "fancy-no-backtrace")] {
+                supports_color::on(supports_color::Stream::Stderr).is_some()
+            } else if #[cfg(feature = "fancy-no-syscall")] {
+                // In no-std environment without color support, default to no color support
```
comment not quite true. fancy-no-syscall can be enabled w/o no-std, right?

huitseeker replied:
amended

cgettys-microsoft commented on src/handler.rs:
```diff
@@ -491,21 +507,27 @@ mod syscall {
     pub(super) fn supports_color() -> bool {
         cfg_if! {
-            if #[cfg(feature = "fancy-no-syscall")] {
+            if #[cfg(feature = "fancy-no-backtrace")] {
```
this reordering seems like it might be a behavior change.
Shouldn't it be something like:
if fancy-no-syscall, false
maybe else if no-std, false
else if fancy-no-backtrace, check
maybe else, false?
These negated features are so hard to think through. Feels like some of them should be flipped next time there's a major version?

cgettys-microsoft commented on src/handler.rs:
```diff
@@ -491,21 +507,27 @@ mod syscall {
-            if #[cfg(feature = "fancy-no-syscall")] {
-                None
-            } else {
+            if #[cfg(feature = "fancy-no-backtrace")] {
```
again, seems like this may change behavior in with std scenarios?

cgettys-microsoft commented on src/handlers/theme.rs:
```diff
@@ -69,12 +71,20 @@ impl GraphicalTheme {
-        match std::env::var("NO_COLOR") {
-            _ if !std::io::stdout().is_terminal() || !std::io::stderr().is_terminal() => {
-                Self::none()
+        #[cfg(feature = "fancy-no-syscall")]
```
this seems ok, though again, the negated features are really tricky...

cgettys-microsoft commented on src/highlighters/mod.rs:
```diff
@@ -78,21 +83,28 @@ impl MietteHighlighter {
-                Self(Arc::new(SyntectHighlighter::default()))
+        #[cfg(all(feature = "syntect-highlighter", not(feature = "fancy-no-syscall")))]
+        {
+            use std::io::IsTerminal;
```
It looks like there's some deduplication that should be done vs src/handlers/theme.rs
It's already pretty complicated and presumably the two should be kept in sync.
Maybe a helper function?  Ideally that memo-izes it to avoid many redundant syscalls if necessary

Note: the double or triple negation in the old version was making my head hurt, so I made it exhaustive instead
```rust
enum StyleDetection
{
  NoColor,
  Color
}
fn detect_color_support() ->
{
    #[cfg(any(feature = "fancy-no-syscall", not(feature = "std")))]
    { 
      StyleDetection::NoColor
    }
    
    #[cfg(all(not(feature = "fancy-no-syscall"), feature = "std"))]
    { 
        match std::env::var("NO_COLOR")  {
          if !std::io::stdout().is_terminal() || !std::io::stderr().is_terminal() => {
            // TODO: address todo by making this StyleDetection::NoColor?
            //TODO: should use ANSI styling instead of 24-bit truecolor here
            StyleDetection::Color
          }
          Ok(string) => 
          {
            if string == "0" 
            {
              // NO_COLOR = "0" means enable color
              StyleDetection::Color
            }
            else
            {
              // any other value means disable color.
              StyleDetection::NoColor
            }
          }
          // This could be NotPresent (color since disablement env var not set) or NotUnicode (hopefully never happens, but we'llt reat it the same)
          Err(_) => StyleDetection::Color
        }
    }
}
```

huitseeker replied:
Thought about this. The two defaults have different return types (`GraphicalTheme` vs `MietteHighlighter`) so a shared helper would need callbacks or an enum. The duplicated logic is small (just the `NO_COLOR` check) and both places already have `cfg` guards. A helper could work but might add more complexity than it saves. Open to ideas if you see a clean way to do it.

cgettys-microsoft commented on src/lib.rs:
```diff
@@ -1,6 +1,8 @@
+#![no_std]
 #![deny(missing_docs, missing_debug_implementations, nonstandard_style)]
```
If this is going to be merged, I think at least some of these should probably be enabled?
```rust
![deny(clippy::alloc_instead_of_core,clippy::std_instead_of_core,clippy::std_instead_of_alloc)]
```

cgettys-microsoft commented on src/lib.rs:
```diff
@@ -815,6 +828,28 @@
+#[cfg(feature = "std")]
+pub use std::error::Error as StdError;
+
+#[cfg(not(feature = "std"))]
```
As written, this isn't additive. Can't there be an enum MietteErr with two enum variants, one of which is only enabled when std is? or would that be breaking? not saying it should change, just seeking to understand

cgettys-microsoft commented on src/protocol.rs:
```diff
@@ -723,6 +729,15 @@ impl SourceOffset {
+    /// as a [`SourceOffset`].
+    ///
+    /// In no_std environments, this is not supported and will return an error.
+    #[cfg(not(feature = "std"))]
```
This doesn't make sense to me, why not just compile it out entirely? why add a useless implementation?

cgettys-microsoft commented on tests/test_boxed.rs:
```diff
@@ -42,13 +42,13 @@ fn test_boxed_str_stderr() {
 
     let error = MyError {
-        source: io::Error::new(io::ErrorKind::Other, "oh no!!!!"),
+        source: io::Error::other("oh no!!!!"),
```
Might be worth pulling all the io::Error changes out into another PR to keep the diff down, but idk.

cgettys-microsoft commented on Cargo.toml:
```diff
@@ -13,9 +13,11 @@ rust-version = "1.82.0"
 exclude = ["images/", "tests/", "miette-derive/"]
 
 [dependencies]
+thiserror = "2.0.11"
```
Why is this needed?

cgettys-microsoft commented on Cargo.toml:
```diff
@@ -13,9 +13,11 @@ rust-version = "1.82.0"
-unicode-width = "0.2.0"
+unicode-width = { version = "0.2.0", default-features = false }
 cfg-if = "1.0.0"
+spin = { version = "0.9", default-features = false, features = ["mutex", "spin_mutex", "lazy"] }
```
Doesn't seem like it should be a dependency for std environments. I see the simplicity benefit, but adding  a dependency for consumers for whom OnceLock is an option, doesn't seem like it makes sense.
Also, do you need all those features?

huitseeker replied:
`spin` is `optional = true` and only activated when `not(feature = "std")` (see the cfg guards in src/eyreish/mod.rs). Std users don't pull in spin at all. The minimal features (once only) are intentional to keep the dependency small for no_std users who need it.

cgettys-microsoft commented on Cargo.toml:
```diff
@@ -13,9 +13,11 @@ rust-version = "1.82.0"
+thiserror = "2.0.11"
 miette-derive = { path = "miette-derive", version = "=7.6.0", optional = true }
-unicode-width = "0.2.0"
+unicode-width = { version = "0.2.0", default-features = false }
```
This is a size optimization? If so, maybe default features should have an option that turns this back on? As written any consumers with cjk character requirements will be broken unless they otherwise include unicode-width with the relevant featur eon

huitseeker replied:
restored

cgettys-microsoft commented on build.rs:
```diff
@@ -0,0 +1,19 @@
+
+fn main() {
+    if let Channel::Nightly = version_meta().unwrap().channel {
+        println!("cargo:rustc-cfg=nightly")
```
Why? the MSRV is way higher than 1.46
Please remove rustc_version too if this goes away

cgettys-microsoft replied:
can't we make the feature on everywhere as well, or is there a reason not to?

cgettys-microsoft replied:
Do we need the build.rs at all, in other words?

cgettys-microsoft commented:
To be clear, my comments are just my opinion - I'm neither the author nor am I a maintainer of this library, I'm just a person who has submitted a few contributions previously and was curious. Hope the feedback is helpful.

huitseeker committed:
fix: address review comments

- Fixed compilation errors in no-std mode (TestError Debug derive, ToString imports)
- Addressed consistency issues: Option qualification, StdError naming, Borrow impl
- Fixed feature flag logic for supports_color function with proper priority ordering
- Added conditional compilation for std-dependent features in panic.rs, handler.rs, theme.rs
- Made syntect-highlighter feature explicitly require std
- Consolidated duplicate Borrow implementations using core::borrow::Borrow

huitseeker committed:
fix: suppress unused assignment warnings in derive tests for Miri

- Add #[allow(unused_assignments)] to test structs/enums with unused fields
- Fix Miri test failures caused by derive macro feature interactions
- Miri tests now pass: 52 passed, 0 failed, 6 ignored
- Issue occurs when fancy feature changes how derive macros process fields
- Suppression is appropriate as fields are only unused under specific feature combinations

cgettys-microsoft commented on Cargo.toml:
```diff
@@ -29,23 +31,23 @@ serde = { version = "1.0.196", features = ["derive"], optional = true }
 serde_json = "1.0.113"
 strip-ansi-escapes = "0.2.0"
 
+[build-dependencies]
```
don't need this any  more, no?

cgettys-microsoft commented on src/handler.rs:
```diff
@@ -366,6 +366,11 @@ impl MietteHandlerOpts {
                     true
                 }
             }
+            #[cfg(all(not(feature = "fancy-no-syscall"), not(feature = "std")))]
```
These two are not opposites. Above is !A && B. this is !A && !B
This should probably be A || !B?

cgettys-microsoft commented on src/panic.rs:
```diff
@@ -1,4 +1,6 @@
+#[cfg(feature = "std")]
 use std::boxed::Box;
+#[cfg(feature = "std")]
 use std::{
```
maybe make the whole module conditional? I don't think a fake no-op set_panic_hook should be provided when std is not set, it's correct to not have it there and will not break any program that depends on miette with std support feature

huitseeker committed:
fix: remove unused rustc_version dependency and simplify build.rs

- Remove rustc_version build dependency that was no longer needed
- Remove nightly detection since nightly cfg flag wasn't used anywhere
- Simplify build.rs to only set track_caller cfg flag

huitseeker committed:
refactor: use #[track_caller] directly instead of conditional cfg_attr

huitseeker committed:
fix: address review feedback for no_std support

- Use core::error::Error directly (MSRV 1.82 supports it)
- Remove build.rs (track_caller is always available with MSRV 1.82)
- Simplify feature detection in handler.rs syscall module
- Add clippy lints to catch std/core/alloc misuse
- Remove custom StdError trait, re-export core::error::Error
- Add __alloc module for derive macro compatibility
- Clean up verbose comments
- Remove extern crate alloc from tests (not needed with std)
- Fix derive macros to use miette::__alloc paths

huitseeker committed:
fix: make spin dependency optional for std builds

- spin is now optional, only needed for no_std (provides Once)
- std builds use OnceLock, avoiding unnecessary dependency
- Restore unicode-width default features (keeps CJK support)

huitseeker commented on src/highlighters/mod.rs:
```diff
@@ -78,21 +83,28 @@ impl MietteHighlighter {
-                Self(Arc::new(SyntectHighlighter::default()))
+        #[cfg(all(feature = "syntect-highlighter", not(feature = "fancy-no-syscall")))]
+        {
+            use std::io::IsTerminal;
```
Thought about this. The two defaults have different return types (`GraphicalTheme` vs `MietteHighlighter`) so a shared helper would need callbacks or an enum. The duplicated logic is small (just the NO_COLOR check) and both places already have cfg guards. A helper could work but might add more complexity than it saves. Open to ideas if you see a clean way to do it.

huitseeker commented on src/highlighters/mod.rs:
```diff
@@ -78,21 +83,22 @@ impl MietteHighlighter {
+        #[cfg(feature = "syntect-highlighter")]
+        {
+            use std::io::IsTerminal;
+            match std::env::var("NO_COLOR") {
```
The two defaults have different return types (`GraphicalTheme` vs `MietteHighlighter`) so a shared helper would need callbacks or an enum. The duplicated logic is small (just the NO_COLOR check) and both places already have cfg guards. A helper could work but might add more complexity than it saves. Open to ideas if you see a clean way to do it.

huitseeker committed:
fix: address review feedback and add feature powerset CI

- Make spin a required dependency for no_std hook support (OnceLock unavailable)
- Simplify cfg guards: not(std) instead of not(std) && spin
- Change clippy alloc/std lints from warn to deny
- Remove no_std stub for from_current_location (compile-time vs runtime error)
- Fix is_graphical cfg logic for fancy-base without fancy-no-backtrace
- Add cargo-hack feature powerset CI job to verify all 56 combinations
